File: src/retrieval/grepRetriever.ts
import fs from "node:fs/promises";
import path from "node:path";
import { spawn } from "node:child_process";
import fg from "fast-glob";
import { type GrepHit } from "../types.js";
import { tokenize } from "../textUtils.js";

export class GrepRetriever {
  constructor(private knowledgeBaseDir: string) {}

  async search(queries: string[], targets: string[], contextLines = 0, maxHitsPerQuery = 80): Promise<GrepHit[]> {
    const batches = await Promise.all(queries.map((query) => this.searchOne(query, targets, contextLines, maxHitsPerQuery)));
    return batches.flat();
  }

  private async searchOne(query: string, targets: string[], contextLines: number, maxHits: number): Promise<GrepHit[]> {
    const terms = tokenize(query);
    if (!terms.length) return [];
    return (await hasRg()) ? this.rg(query, terms, targets, contextLines, maxHits) : this.nodeSearch(query, terms, targets, maxHits);
  }

  private async rg(query: string, terms: string[], targets: string[], contextLines: number, maxHits: number): Promise<GrepHit[]> {
    const pattern = terms.join("|");
    const args = ["--line-number", "--ignore-case", "--no-heading", "--glob", "*.md", "-C", String(contextLines), pattern, ...targets];
    const stdout = await run("rg", args);
    const hits: GrepHit[] = [];
    for (const raw of stdout.split(/\r?\n/)) {
      const parsed = parseRgLine(raw);
      if (!parsed) continue;
      const matchedTerms = terms.filter((term) => parsed.lineText.toLowerCase().includes(term.toLowerCase()));
      if (matchedTerms.length) {
        hits.push({
          file: relativeTo(this.knowledgeBaseDir, parsed.file),
          lineNumber: parsed.lineNumber,
          lineText: parsed.lineText,
          query,
          matchedTerms
        });
      }
      if (hits.length >= maxHits) break;
    }
    return hits;
  }

  private async nodeSearch(query: string, terms: string[], targets: string[], maxHits: number): Promise<GrepHit[]> {
    const files = (await Promise.all(targets.map((target) => markdownFiles(target)))).flat();
    const hits: GrepHit[] = [];
    for (const file of files) {
      const lines = (await fs.readFile(file, "utf8")).split(/\r?\n/);
      lines.forEach((lineText, idx) => {
        const matchedTerms = terms.filter((term) => lineText.toLowerCase().includes(term.toLowerCase()));
        if (matchedTerms.length && hits.length < maxHits) {
          hits.push({ file: relativeTo(this.knowledgeBaseDir, file), lineNumber: idx + 1, lineText, query, matchedTerms });
        }
      });
      if (hits.length >= maxHits) break;
    }
    return hits;
  }
}

async function markdownFiles(target: string): Promise<string[]> {
  const stat = await fs.stat(target);
  if (stat.isFile() && target.endsWith(".md")) return [target];
  if (stat.isDirectory()) return fg("**/*.md", { cwd: target, onlyFiles: true, absolute: true });
  return [];
}

async function hasRg(): Promise<boolean> {
  try {
    await run("rg", ["--version"]);
    return true;
  } catch {
    return false;
  }
}

function run(command: string, args: string[]): Promise<string> {
  return new Promise((resolve, reject) => {
    const child = spawn(command, args, { stdio: ["ignore", "pipe", "ignore"] });
    let stdout = "";
    child.stdout.on("data", (chunk) => (stdout += String(chunk)));
    child.on("error", reject);
    child.on("close", (code) => (code === 0 || code === 1 ? resolve(stdout) : reject(new Error(`${command} exited ${code}`))));
  });
}

function parseRgLine(raw: string): { file: string; lineNumber: number; lineText: string } | undefined {
  const separator = raw.includes(":") ? ":" : "-";
  const parts = raw.split(separator);
  if (parts.length < 3) return undefined;
  const file = parts[0] ?? "";
  const lineNumber = Number(parts[1]);
  const lineText = parts.slice(2).join(separator);
  return Number.isFinite(lineNumber) ? { file, lineNumber, lineText } : undefined;
}

function relativeTo(root: string, file: string): string {
  const relative = path.relative(path.resolve(root), path.resolve(file));
  return relative.startsWith("..") ? file : relative;
}

File: app/retrieval/grep_retriever.py
from __future__ import annotations

import asyncio
import shutil
from pathlib import Path

from app.schemas.retrieval import GrepHit
from app.text_utils import tokenize


class GrepRetriever:
    def __init__(self, knowledge_base_dir: Path):
        self.knowledge_base_dir = knowledge_base_dir
        self.has_rg = shutil.which("rg") is not None

    async def search(
        self,
        queries: list[str],
        targets: list[Path],
        context_lines: int = 0,
        max_hits_per_query: int = 80,
    ) -> list[GrepHit]:
        if not targets or not queries:
            return []
        tasks = [
            self._search_one(query, targets, context_lines, max_hits_per_query)
            for query in queries
        ]
        results = await asyncio.gather(*tasks)
        return [hit for batch in results for hit in batch]

    async def _search_one(
        self,
        query: str,
        targets: list[Path],
        context_lines: int,
        max_hits: int,
    ) -> list[GrepHit]:
        terms = tokenize(query)
        if not terms:
            return []
        if self.has_rg:
            return await self._rg(query, terms, targets, context_lines, max_hits)
        return await asyncio.to_thread(self._python_search, query, terms, targets, max_hits)

    async def _rg(
        self,
        query: str,
        terms: list[str],
        targets: list[Path],
        context_lines: int,
        max_hits: int,
    ) -> list[GrepHit]:
        pattern = "|".join(terms)
        cmd = [
            "rg",
            "--line-number",
            "--ignore-case",
            "--no-heading",
            "--glob",
            "*.md",
            "-C",
            str(context_lines),
            pattern,
            *[str(target) for target in targets],
        ]
        proc = await asyncio.create_subprocess_exec(
            *cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.DEVNULL,
        )
        stdout, _ = await proc.communicate()
        hits: list[GrepHit] = []
        for raw in stdout.decode(errors="replace").splitlines():
            parsed = self._parse_rg_line(raw)
            if not parsed:
                continue
            file, line_number, line_text = parsed
            matched = [term for term in terms if term.lower() in line_text.lower()]
            if matched:
                hits.append(
                    GrepHit(
                        file=self._relative(file),
                        line_number=line_number,
                        line_text=line_text,
                        query=query,
                        matched_terms=matched,
                    )
                )
            if len(hits) >= max_hits:
                break
        return hits

    def _python_search(self, query: str, terms: list[str], targets: list[Path], max_hits: int) -> list[GrepHit]:
        files: list[Path] = []
        for target in targets:
            if target.is_file() and target.suffix == ".md":
                files.append(target)
            elif target.is_dir():
                files.extend(target.rglob("*.md"))
        hits: list[GrepHit] = []
        for path in files:
            try:
                lines = path.read_text(encoding="utf-8").splitlines()
            except UnicodeDecodeError:
                lines = path.read_text(errors="replace").splitlines()
            for idx, line in enumerate(lines, start=1):
                matched = [term for term in terms if term.lower() in line.lower()]
                if matched:
                    hits.append(
                        GrepHit(
                            file=self._relative(path),
                            line_number=idx,
                            line_text=line,
                            query=query,
                            matched_terms=matched,
                        )
                    )
                if len(hits) >= max_hits:
                    return hits
        return hits

    def _parse_rg_line(self, raw: str) -> tuple[Path, int, str] | None:
        separator = ":" if ":" in raw else "-"
        parts = raw.split(separator, 2)
        if len(parts) != 3:
            return None
        file_raw, line_raw, line_text = parts
        try:
            line_number = int(line_raw)
        except ValueError:
            return None
        return Path(file_raw), line_number, line_text

    def _relative(self, path: Path) -> str:
        try:
            return str(path.resolve().relative_to(self.knowledge_base_dir.resolve()))
        except ValueError:
            return str(path)

